File: app/src/main/java/edu/gatech/spellastory/data/Categories.java
package edu.gatech.spellastory.data;

import com.opencsv.CSVReader;

import java.io.IOException;
import java.io.Reader;
import java.util.ArrayList;
import java.util.HashMap;
import java.util.List;
import java.util.Map;

import edu.gatech.spellastory.domain.Category;

public class Categories {

    // Maps category # to category name
    private Map<Integer, String> categoryMap;

    public Categories(Reader reader) throws IOException {
        categoryMap = readCategories(reader);
    }

    private Map<Integer, String> readCategories(Reader reader) throws IOException {
        Map<Integer, String> categories = new HashMap<>();
        CSVReader r = new CSVReader(reader);
        String[] nextLine;
        while ((nextLine = r.readNext()) != null) {
            int num = Integer.parseInt(nextLine[0]);
            String name = nextLine[1];
            categoryMap.put(num, name);
        }
        return categories;
    }

    public Category getCategory(int code) {
        String name = categoryMap.get(code);
        return new Category(code, name);
    }

    public List<Category> getCategories(String... codes) {
        int[] intCodes = new int[codes.length];
        for (int i = 0; i < codes.length; i++) {
            intCodes[i] = Integer.parseInt(codes[i]);
        }

        return getCategories(intCodes);
    }

    public List<Category> getCategories(int... codes) {
        List<Category> categories = new ArrayList<>(codes.length);
        for (int code : codes) {
            categories.add(getCategory(code));
        }
        return categories;
    }
}

File: app/src/main/java/edu/gatech/spellastory/data/Categories.kt
package edu.gatech.spellastory.data

import android.content.Context
import com.opencsv.CSVIterator
import com.opencsv.CSVReader
import edu.gatech.spellastory.domain.Category
import edu.gatech.spellastory.util.SingletonHolder
import java.io.InputStreamReader

class Categories private constructor(context: Context) {

    private val categories: Map<Int, String> by lazy {
        val reader = CSVReader(InputStreamReader(context.assets.open("coded_categories.csv")))
        val categories = read(reader)
        reader.close()
        categories
    }

    private fun read(reader: CSVReader) = CSVIterator(reader).asSequence().map { line ->
        val num = Integer.parseInt(line[0])
        val name = line[1]
        num to name
    }.toMap()

    fun getCategory(code: Int): Category = Category(code, categories.getValue(code))

    fun getCategories(vararg codes: String): List<Category> =
        getCategories(*codes.map { Integer.parseInt(it) }.toIntArray())

    fun getCategories(vararg codes: Int): List<Category> = codes.map { getCategory(it) }

    companion object : SingletonHolder<Categories, Context>(::Categories)
}